experiment/clustersecretbackup/main.go

```diff
@@ -0,0 +1,256 @@
+}
+
+// gatherOptions parses the command-line flags.
+func gatherOptions(fs *flag.FlagSet, args ...string) options {
+	var o options
+	fs.StringVar(&o.project, "project", "", "GCP project used for backing up secrets")
+	fs.StringVar(&o.cluster, "cluster", "", "cluster context name used for backing up secrets")
+	fs.Var(&o.namespaces, "namespace", "namespace to backup, can be passed in repeatedly")
+	fs.BoolVar(&o.update, "update", false, "Controls whether update existing secret or not, if false then secret will only be created")
+	fs.BoolVar(&o.dryRun, "dryrun", false, "Controls whether this is dry run or not")
+	fs.Parse(args)
+
+	return o
+}
+
+// validateFlags validates the command-line flags.
+func (o *options) validateFlags() error {
+	if len(o.project) == 0 {
+		return errors.New("--project must be provided")
+	}
+	if len(o.cluster) == 0 {
+		return errors.New("--cluster must be provided")
+	}
+	return nil
+}
+
+func newClient(o options) (*client, error) {
+	secretmanagerClient, err := secretmanager.NewClient(o.project)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating secret manager client: %w", err)
+	}
+
+	kubeClient, err := newKubeClients("", o.cluster)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating kube client: %w", err)
+	}
+	return &client{
+		kubeClient:          kubeClient,
+		secretmanagerClient: secretmanagerClient,
+		options:             o,
+	}, nil
+}
+
+func newKubeClients(configPath, clusterName string) (ctrlruntimeclient.Client, error) {
+	var loader clientcmd.ClientConfigLoader
+	if configPath != "" {
+		loader = &clientcmd.ClientConfigLoadingRules{ExplicitPath: configPath}
+	} else {
+		loader = clientcmd.NewDefaultClientConfigLoadingRules()
+	}
+
+	overrides := clientcmd.ConfigOverrides{}
+	// Override the cluster name if provided.
+	if clusterName != "" {
+		overrides.Context.Cluster = clusterName
+		overrides.CurrentContext = clusterName
+	}
+
+	cfg, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		loader, &overrides).ClientConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed create rest config: %v", err)
+	}
+	return ctrlruntimeclient.New(cfg, ctrlruntimeclient.Options{})
+}
+
+func backupSecret(ctx context.Context, secretmanagerClient secretmanager.ClientInterface, secretID string, payload []byte) error {
+	ss, err := secretmanagerClient.ListSecrets(ctx)
+	if err != nil {
+		return err
+	}
+	var found bool
+	for _, s := range ss {
+		if strings.HasSuffix(s.Name, fmt.Sprintf("/%s", secretID)) {
+			found = true
+		}
+	}
+	if !found {
+		if _, err = secretmanagerClient.CreateSecret(ctx, secretID); err != nil {
+			return err
+		}
+	}
+	return secretmanagerClient.AddSecretVersion(ctx, secretID, payload)
+}
+
+// process merges secret into a new secret for write.
+func (c *client) updateSingleSecret(ctx context.Context, si *secretInfo) error {
+	secretID := si.gsmSecretName()
+	payload, err := yaml.Marshal(si.clusterSecret)
```

Comment:
Good point, done and converted this to json so that it can be processed by kubernetes external manager later. Have tested it manually and the generated secret could be used for this purpose